Sources/NIO/ThreadPosix.swift

```diff
@@ -0,0 +1,150 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the SwiftNIO open source project
+//
+// Copyright (c) 2020 Apple Inc. and the SwiftNIO project authors
+// Licensed under Apache License v2.0
+//
+// See LICENSE.txt for license information
+// See CONTRIBUTORS.txt for the list of SwiftNIO project authors
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+//===----------------------------------------------------------------------===//
+
+#if !os(Windows)
+
+
+#if os(Linux)
+import CNIOLinux
+
+private let sys_pthread_getname_np = CNIOLinux_pthread_getname_np
+private let sys_pthread_setname_np = CNIOLinux_pthread_setname_np
+private typealias ThreadDestructor = @convention(c) (UnsafeMutableRawPointer?) -> UnsafeMutableRawPointer?
+#elseif os(iOS) || os(macOS) || os(tvOS) || os(watchOS)
+private let sys_pthread_getname_np = pthread_getname_np
+// Emulate the same method signature as pthread_setname_np on Linux.
+private func sys_pthread_setname_np(_ p: pthread_t, _ pointer: UnsafePointer<Int8>) -> Int32 {
+    assert(pthread_equal(pthread_self(), p) != 0)
+    pthread_setname_np(pointer)
+    // Will never fail on macOS so just return 0 which will be used on linux to signal it not failed.
+    return 0
+}
+private typealias ThreadDestructor = @convention(c) (UnsafeMutableRawPointer) -> UnsafeMutableRawPointer?
+
+#endif
```

Comment:
Instead of nesting in `#if !os(Windows)` can we just have a separate `#if os(Linux) || os(iOS) || os(macOS) || os(tvOS) || os(watchOS)` below here; it's closer to what we actually want, right?